### PR TITLE
util, executor: use a individual rowContainer for each probeWorker in HashJoinExec

### LINKED ISSUE #27576: bottleneck when accessing rowContainer.m.Mutex in HashJoinExec
### Workload
[1W.txt](https://github.com/pingcap/tidb/files/7044518/1W.txt)
[150W.txt](https://github.com/pingcap/tidb/files/7044519/150W.txt)

``` sql
drop table if exists t1, t2, t3;
create table t1(c1 int);
create table t2(c1 int);
create table t3(c1 int);
load data local infile '$dir$/150W.sql' into table t1 fields terminated by ' ';
load data local infile '$dir$/1W.sql' into table t2 fields terminated by ' ';
load data local infile '$dir$/1W.sql' into table t3 fields terminated by ' ';
```

### Test SQL
``` sql
set @@tidb_executor_concurrency=$N$;
select /*+ hash_join(t1, t2) */ * from t1 inner join t2 inner join t3 on t1.c1 = t2.c1 and t2.c1 = t3.c1 limit 1 offset 153599999;
```

### Test Result

|tidb_executor_concurrency|latency|
|--|--|
|2|25.78s|
|4|17.14s|
|8|17.30s|
|16|17.76s|
|32|17.38s|

### Analysis
The test result shows that the query performance does not improve as the concurrency increase.
After the profile, we can see that there is a bottleneck when invoking `rowContainer.GetRow`.
截屏2021-08-25 下午3 25 27

### Summary
We can refine the usage of Mutex in RowContainer to eliminate the bottleneck.

### PR DESCRIPTION
**Pull request by XuHuaiyu**
### What problem does this PR solve?

Issue Number: close #27576 

Problem Summary:

### What is changed and how it works?

Proposal: [xxx](url) 

What's Changed:
Use an individual rowContainer for each probeWorker in HashJoinExec to eliminate the contention when accessing rowContainer.Mutex.

How it Works:

### Check List 

Tests 

- [ ] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)

|tidb_executor_concurrency|lantency|
|--|--|
|2|20.78s|
|4|13.93s|
|8|8.18s|
|16|5.56s|
|32|3.42s|

截屏2021-08-25 下午3 32 35

Side effects

N/A

Documentation

N/A

### Release note 

```release-note
Improve the query performance for HashJoin.
```

### CODE REVIEW

**Comment by XuHuaiyu** [comment]
The failed cases: [TestShowGrantsForCurrentUserUsingRole](https://github.com/pingcap/tidb/issues/27651) and [TestIterator](https://github.com/pingcap/tidb/issues/26235) have nothing to do with this commit.

File: util/chunk/row_container.go [hunk 1]
```diff
@@ -28,19 +28,55 @@ import (
...
+	// goroutines touch the same rowContainer concurrently, it's recommended to
+	// use RowContainer.ShallowCopyWithNewMutex to build a new RowContainer for
+	// each goroutine. Thus each goroutine holds its own rLock but share the same
+	// underlying data, which can reduce the contention on m.rLock remarkably and
+	// get better performance.
+	rLock   *sync.RWMutex
+	wLocks  []*sync.RWMutex
+	records *rowContainerRecord
+	// spillError stores the error when spilling.
+	spillError error
```
[the remark this answers is not in the document]
Comment (XuHuaiyu) [reply util/chunk/row_container.go:49]: `spillError` is of the `interface` type, it seems we do not need to put it in rowContainerRecord

File: executor/hash_table.go [hunk 2]
```diff
@@ -115,7 +125,7 @@ func (c *hashRowContainer) GetMatchedRowsAndPtrs(probeKey uint64, probeRow chunk
 			return
 		}
 		if !ok {
-			c.stat.probeCollision++
+			atomic.AddInt64(&c.stat.probeCollision, 1)
```
Comment (guo-shaoge) [inline executor/hash_table.go:128]: Guess this can be a new hotspot. Myabe we can let user control the HashTalbe's bucket size to fix.

**Review by guo-shaoge — APPROVED** [review]

File: executor/hash_table.go (hunk 2 above)
Comment (XuHuaiyu) [reply executor/hash_table.go:128]: Actually, the collision should be small probability event

**Review by wshwsh12 — APPROVED** [review]
LGTM

**Pull request merged**